livekit-webrtc/src/native/audio_source.rs

```diff
@@ -73,52 +76,195 @@ impl NativeAudioSource {
-                    inner.offset += remaining_data;
-                }
+    pub fn sample_rate(&self) -> u32 {
+        self.sample_rate
+    }
 
-                break;
-            }
+    pub fn num_channels(&self) -> u32 {
+        self.num_channels
+    }
 
-            let data = if inner.offset != 0 {
-                // Use the data from the previous capture
-                let data = &mut inner.buf[buf_offset..];
-                data.copy_from_slice(&frame.data[i..i + needed_data]);
-                inner.offset = 0;
+    // Implemented inside another functions to allow unit testing
+    fn next_frame<'a>(
+        &self,
+        inner: &'a mut MutexGuard<'_, AudioSourceInner>, // The lock musts be guarded by capture_frame
+        frame: &'a AudioFrame<'a>,
+    ) -> Option<&'a [i16]> {
+        let available_data = inner.len + frame.data.len() - inner.read_offset;
+        if available_data >= self.samples_10ms {
+            // Capture the frame to libwebrtc
+            Some(if inner.len != 0 {
+                // Read 10ms frame from inner.buf AND frame.data
+                let missing_len = self.samples_10ms - inner.len;
+                let start = inner.len;
+                inner.buf[start..].copy_from_slice(&frame.data[..missing_len]);
+                inner.read_offset += missing_len;
+                inner.len = 0;
                 &inner.buf
             } else {
-                &frame.data[i..i + samples_10ms]
+                // Read 10ms frame only from frame.data
+                let start = inner.read_offset;
+                let end = start + self.samples_10ms;
+                inner.read_offset += self.samples_10ms;
+                &frame.data[start..end]
+            })
+        } else {
+            // Save to buf and wait for the next capture_frame to give enough data to complete a 10ms frame
+            let remaining_data = frame.data.len() - inner.read_offset; // remaining data from frame.data
+            let start = inner.len;
+            let end = start + remaining_data;
+            let start2 = frame.data.len() - remaining_data;
+            inner.buf[start..end].copy_from_slice(&frame.data[start2..]);
+            inner.len += remaining_data;
+            inner.read_offset = 0;
+            None
+        }
+    }
+
+    pub async fn capture_frame(&self, frame: &AudioFrame<'_>) -> Result<(), RtcError> {
+        if self.sample_rate != frame.sample_rate || self.num_channels != frame.num_channels {
+            return Err(RtcError {
+                error_type: RtcErrorType::InvalidState,
+                message: "sample_rate and num_channels don't match".to_owned(),
+            });
+        }
+
+        let mut inner = self.inner.lock().await;
+        let mut interval = interval(Duration::from_millis(10));
+
+        loop {
+            let Some(data) = self.next_frame(&mut inner, frame) else {
+                break;
             };
 
+            interval.tick().await;
```

Comment:
would there be clock drift issues here? i.e.
* 100ns spent in processing (between on_captured_frame and next_frame)
* sleeps for 10ms
* repeat 10 times, audio would be behind by 1ms

Reply:
So I now keep the same intervals between capture_frame calls, I think the tokio timer precision should be good enough for 10ms

Reply:
sorry, I misunderstood the `tick` part. understood now that you are keeping the same interval obj.